stratum/hal/lib/tdi/es2k/es2k_sde_packetio.cc

```diff
@@ -0,0 +1,428 @@
+    printf("%02X", raw_data[j]);
+    if (j < 11) printf(":");
+  }
+  printf("\n");
+
+  printf("pktio: Dst MAC: ");
+  for (int j = 0; j < 6; ++j) {
+    printf("%02X", raw_data[j]);
+    if (j < 5) printf(":");
+  }
+  printf("\n");
+
+  inet_ntop(AF_INET, &(ip_hdr->ip_src), src_ip_str, INET_ADDRSTRLEN);
+  inet_ntop(AF_INET, &(ip_hdr->ip_dst), dst_ip_str, INET_ADDRSTRLEN);
+
+  printf("pktio: Src IP: %s\n", src_ip_str);
+  printf("pktio: Dst IP: %s\n", dst_ip_str);
+}
+
+::util::Status Es2kSdeWrapper::RegisterPacketReceiveWriter(
+    int device, std::unique_ptr<ChannelWriter<std::string>> writer) {
+  absl::WriterMutexLock l(&packet_rx_callback_lock_);
+  device_to_packet_rx_writer_[device] = std::move(writer);
+  return ::util::OkStatus();
+}
+
+::util::Status Es2kSdeWrapper::UnregisterPacketReceiveWriter(int device) {
+  absl::WriterMutexLock l(&packet_rx_callback_lock_);
+  device_to_packet_rx_writer_.erase(device);
+  return ::util::OkStatus();
+}
+
+::util::Status Es2kSdeWrapper::SetPacketIoConfig(
+    const PacketIoConfig& pktio_config) {
+  pktio_config_.CopyFrom(pktio_config);
+  return ::util::OkStatus();
+}
+
+// TxCallback that is called by sde for Tx done.
+void Es2kSdeWrapper::PktIoTxCallback(
+    std::unique_ptr<::tdi::TableKey> key,
+    std::unique_ptr<::tdi::TableData> data,
+    std::unique_ptr<::tdi::NotificationParams> pkts_info_recv, void* cookie) {
+  uint64_t port_id;
+  uint64_t queue_id;
+  uint64_t nb_pkts;
+  std::vector<uint64_t> pkt_len;  /* len of pkts */
+  std::vector<uint64_t> pkt_data; /* pointers to pkts data */
+
+  // parse callback params
+  pkts_info_recv->getValue(PORT_ID, &port_id);
+  pkts_info_recv->getValue(QUEUE_ID, &queue_id);
+  pkts_info_recv->getValue(NB_PKTS, &nb_pkts);
+  pkts_info_recv->getValue(PKT_LEN, &pkt_len);
+  pkts_info_recv->getValue(PKT_DATA, &pkt_data);
+
+  printf(
+      "[satish] pktio: callbacks received for %lu pkts sent on port_id: %lu, "
+      "queue_id: %lu\n",
+      nb_pkts, port_id, queue_id);
+
+  // freeing the buffers that were allocated during Tx
+  for (uint64_t i = 0; i < nb_pkts; i++) {
+    printf("[satish] Freeing: %p\n", reinterpret_cast<uint8_t*>(pkt_data[i]));
+    delete[] reinterpret_cast<uint8_t*>(pkt_data[i]);
+  }
+}
+
+// TxPacket involves three steps
+// 1. Allocate operations of type transmit_pkt
+// 2. Fill operations object with packet's data
+// 3. Transmit the packet by calling OperationExecute
+::util::Status Es2kSdeWrapper::TxPacket(int dev_id, const std::string& buffer) {
+  if (pktio_config_.ports_size() == 0) {
+    LOG(INFO) << "pktio not configured, can't transmit packet";
+    std::cout << "[satish] pktio not configured, can't transmit packet"
+              << std::endl;
+    return ::util::OkStatus();
+  }
+
+  // create Target
+  const ::tdi::Device* device = nullptr;
+  ::tdi::DevMgr::getInstance().deviceGet(dev_id, &device);
+  std::unique_ptr<::tdi::Target> dev_tgt;
+  device->createTarget(&dev_tgt);
+
+  // get table
```

Comment:
```suggestion
  // Get pointer to packet i/o table.
```

"Just enough information"

Reply:
Done.